### ISSUE TITLE
ampgo evaluating function out of its scope

### ISSUE THREAD

**Issue by TiagoBsCastro** [lmfit/lmfit-py#602]
Hi guys,

   I have been using lmfit.minimize with ampgo algorithm in order to find the best fit for a function. As it failed several times returning me that the input had NaN values I wrote the following iter_cb:

def iter_cb_HMF(params, iter, resid):
    
    if resid == np.nan or resid == np.inf:
        v = params.valuesdict()
        print iter, [v['A0'],v['Az'],v['L0'],v['k0'],v['s0'],v['a0'],v['az'],v['b0'],v['bz'],v['c0'],v['cz']], resid

that prints the parameters in case the residual is nan or inf. Running it again the code it returned me:

mi = lmfit.minimize(residual_HMF, p_HMF, method='ampgo',iter_cb=iter_cb_HMF)
lmfit.printfuncs.report_fit(mi.params, min_correl=0.5)

316057 [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan] inf

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-36-cbb05fdb10a5> in <module>()
----> 1 mi = lmfit.minimize(residual_HMF, p_HMF, method='ampgo',iter_cb=iter_cb_HMF)
      2 lmfit.printfuncs.report_fit(mi.params, min_correl=0.5)

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/minimizer.pyc in minimize(fcn, params, method, args, kws, iter_cb, scale_covar, nan_policy, reduce_fcn, calc_covar, **fit_kws)
   2446                        nan_policy=nan_policy, reduce_fcn=reduce_fcn,
   2447                        calc_covar=calc_covar, **fit_kws)
-> 2448     return fitter.minimize(method=method)

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/minimizer.pyc in minimize(self, method, params, **kws)
   2115                         val.lower().startswith(user_method)):
   2116                     kwargs['method'] = val
-> 2117         return function(**kwargs)
   2118 
   2119 

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/minimizer.pyc in ampgo(self, params, **kws)
   1853 
   1854         try:
-> 1855             ret = ampgo(self.penalty, values, **ampgo_kws)
   1856         except AbortFitException:
   1857             pass

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/_ampgo.pyc in ampgo(objfun, x0, args, local, local_opts, bounds, maxfunevals, totaliter, maxiter, glbtol, eps1, eps2, tabulistsize, tabustrategy, disp)
    200 
    201             res = minimize(tunnel, x0, args=tunnel_args, method=local,
--> 202                            bounds=bounds, tol=local_tol, options=options)
    203             xf, yf, num_fun = res['x'], res['fun'], res['nfev']
    204             if isinstance(yf, np.ndarray):

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/scipy/optimize/_minimize.pyc in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    599     elif meth == 'l-bfgs-b':
    600         return _minimize_lbfgsb(fun, x0, args, jac, bounds,
--> 601                                 callback=callback, **options)
    602     elif meth == 'tnc':
    603         return _minimize_tnc(fun, x0, args, jac, bounds, callback=callback,

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/scipy/optimize/lbfgsb.pyc in _minimize_lbfgsb(fun, x0, args, jac, bounds, disp, maxcor, ftol, gtol, eps, maxfun, maxiter, iprint, callback, maxls, **unknown_options)
    333             # until the completion of the current minimization iteration.
    334             # Overwrite f and g:
--> 335             f, g = func_and_grad(x)
    336         elif task_str.startswith(b'NEW_X'):
    337             # new iteration

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/scipy/optimize/lbfgsb.pyc in func_and_grad(x)
    278     if jac is None:
    279         def func_and_grad(x):
--> 280             f = fun(x, *args)
    281             g = _approx_fprime_helper(x, fun, epsilon, args=args, f0=f)
    282             return f, g

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/scipy/optimize/optimize.pyc in function_wrapper(*wrapper_args)
    298     def function_wrapper(*wrapper_args):
    299         ncalls[0] += 1
--> 300         return function(*(wrapper_args + args))
    301 
    302     return ncalls, function_wrapper

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/_ampgo.pyc in tunnel(x0, *args)
    272         fun_args = tuple(args[3:])
    273 
--> 274     numerator = (objfun(x0, *fun_args) - aspiration)**2
    275     denominator = 1.0
    276 

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/minimizer.pyc in penalty(self, fvars)
    599             apply_bounds_transformation = True
    600 
--> 601         r = self.__residual(fvars, apply_bounds_transformation)
    602         if isinstance(r, ndarray) and r.size > 1:
    603             r = self.reduce_fcn(r)

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/minimizer.pyc in __residual(self, fvars, apply_bounds_transformation)
    546         else:
    547             return _nan_policy(np.asarray(out).ravel(),
--> 548                                nan_policy=self.nan_policy)
    549 
    550     def __jacobian(self, fvars):

/home/tcastro/anaconda3/envs/base2/lib/python2.7/site-packages/lmfit/minimizer.pyc in _nan_policy(arr, nan_policy, handle_inf)
   2310                    'handle this! Please read https://lmfit.github.io/lmfit-py/faq.html#i-get-errors-from-nan-in-my-fit-what-can-i-do '
   2311                    'for more information.')
-> 2312             raise ValueError(msg)
   2313     return arr
   2314 

ValueError: NaN values detected in your input data or the output of your objective/model function - fitting algorithms cannot handle this! Please read https://lmfit.github.io/lmfit-py/faq.html#i-get-errors-from-nan-in-my-fit-what-can-i-do for more information.

As you can see the residual is inf because the parameters are going crazy. How come can this happen? 

To further test that my function is not going bananas as well, I have run the same thing with 'dual_annealing' and it runs just fine:

mi = lmfit.minimize(residual_HMF, p_HMF, method='dual_annealing',iter_cb=None)
lmfit.printfuncs.report_fit(mi.params, min_correl=0.5)
executed in 3m 46s, finished 10:33:57 2019-11-13
[[Variables]]
    A0:  0.62615107 (init = 0.3536428)
    Az: -0.04742277 (init = -0.0280328)
    L0:  0.47131479 (init = 0.1)
    k0: -0.16228710 (init = -0.01)
    s0:  0.82731502 (init = 1)
    a0:  1.26943759 (init = 2.214875)
    az:  0.46150161 (init = -0.1108433)
    b0:  0.95915886 (init = 1.307311)
    bz:  0.23225713 (init = -0.1623317)
    c0:  1.16079602 (init = 1.284506)
    cz:  0.12335394 (init = -0.05362131)

I'm using:
Python: 2.7.15 |Anaconda, Inc.| (default, Oct 10 2018, 21:32:13) [GCC 7.3.0]
lmfit: 0.9.14, scipy: 1.2.1, numpy: 1.16.5, asteval: 0.9.15, uncertainties: 3.1.2, six: 1.12.0

I warmly appreciate your attention and please do not hesitate in asking further details

**Comment by reneeotten** [comment]
this likely should have been a question on the mailinglist instead of a GitHub Issue.... 

In both cases though the instructions clearly tell you to give a "minimal, verifiable, working example" to reproduce the issue, which you didn't do. ```lmfit``` tells you that you have "NaN values detected in your input data or the output of your objective/model function", but you don't actually show us your input data nor the objective function. 

It's very well possible that there is actually a bug in how I implemented ```ampgo```, but with the information you provided that's hard to tell; so for now this looks much more like a question.

**Comment by TiagoBsCastro** [comment]
Well, I apologize for the misunderstanding, however, as it seems that the algorithm is passing crazy values to the function and not the other way around I don't see it as a simple question. 

I understand that passing a minimal, verifiable, working example helps to debug. However, given the complexity of my code, it is not possible to provide that in the near future. I am sorry.

I would nevertheless check how the implementation is generating new parameters. My guess is that the tunneling function is trying to divide by 0 or something like that.

I again apologize if it is not helpful at all and understand if you want to close it.

**Comment by newville** [comment]
@TiagoBsCastro  Well, if you read all the advice and warning about raising an github issue vs posting a question on the mailing list and decided that what you were seeing is an issue with the code, then you are sort of obligated to do the work to provide a minimal verifiable example that shows the problem.  You did not do this.

So, you *claim* there is a problem with the code.  But you do not *show* there is a problem.  How could you possibly expect us to solve such a problem?   Again, you explicitly and deliberately decided that you were not asking for help getting your code to work with possible suspicion that there might be a problem with `ampgo`.  Nope, you are saying there is a problem with `ampgo`, and not with your code.  And you decided, now twice, to give no evidence of that.

I hope you now see why it is better start with a question of why something isn't working for you, because clearly reading all the advice about this and @reneeotten reply to your issue was not enough.    So, in short and without mincing words: post an example to convince us that the error is not with your code.

**Comment by TiagoBsCastro** [comment]
Dear @newville and @reneeotten it was really not my intention to seem snob. If my text seemed like this it was due to my lack of fluency in English.

In addition, if opening the issue and replying gave you the idea that I was unrespectful to you or to your work I again deeply apologize. Lmfit is one of my favorite packages and I had no intention but help.

I opened the issue because I understood that the test that I have done (iter_cb returning me nan values) was definitive prooving that something internally was not working well --- as this did not convince you, I imagine that there are other possibilities that my limited understanding of both the method and programming does not allow me to see.  I will then close the issue and open this discussion as you guys suggested. Sorry for this and thank you for your attention.
